### ISSUE TITLE
MockDataProvider cannot handle MULTISET fields

### ISSUE THREAD

**Issue by iamtakingiteasy** [jOOQ/jOOQ#12308]
How mocking of multisets is supposed to work is unclear. 

Following examples from https://www.jooq.org/doc/3.7/manual/tools/jdbc-mocking/

```java
var attributes = DSL.multiset(
  DSL.selectFrom(NESTED_TABLE)
).as("attributes");

var fields = TABLE.fieldStream().collect(Collectors.toList());
fields.add(attributes);

var create = DSL.using(SQLDialect.POSTGRES);
MockResult[] mock = new MockResult[1];

var result = create.newResult(fields);

var rec = create.newRecord(fields);
rec.setValue(attributes, create.newResult(NESTED_TABLE));

result.add(rec);

mock[0] = new MockResult(1, result);
```

Looks fine, but in run-time `org.jooq.impl.DefaultBinding.DefaultResultBinding#readMultiset` attempts to parse string representation of  Result as JSON, while there also isn't an obvious way to set JSON representation to `rec`'s "attributes" field (all set* methods are strongly typed), nor a way to avoid deserialization attempts (readMultiset implementation/usage doesn't suggest any alternatives).

I suppose one could reflect his way into AbstractRecord, and indeed something like
```java
try {
    Method method = Class.forName("org.jooq.impl.AbstractRecord").getDeclaredMethod("set", int.class, Field.class, Object.class);
    method.setAccessible(true);
    method.invoke(rec, rec.indexOf(attributes), attributes, "[]");
} catch (Exception e) {
    throw new RuntimeException(e);
}
```
would work, but that probably not an intended way of jOOQ MockDataProvider API.

### Versions:
- jOOQ: 3.15.1
- Java: 11
- Database (include vendor): postgres 12 / mocked
- OS: *nix
- JDBC Driver (include name if inofficial driver): none / mock

**Comment by lukaseder** [comment]
Thanks for your report. Looks like a bug. There's no reason why these things shouldn't work out of the box.

**Comment by lukaseder** [comment]
This test reproduces it:

```java
@Test
public void testMultiset() {
    Field<Result<Table1Record>> ms = multiset(selectFrom(TABLE1));

    DSLContext c = DSL.using(POSTGRES);
    Record1<Result<Table1Record>> r = c.newRecord(ms);
    r.values(resultOne);

    for (NestedCollectionEmulation e : Arrays.asList(
        NestedCollectionEmulation.DEFAULT,
        NestedCollectionEmulation.JSON,
        NestedCollectionEmulation.JSONB,
        NestedCollectionEmulation.XML
    )) {
        DSLContext ctx = DSL.using(new MockConnection(Mock.of(r)), POSTGRES, new Settings().withEmulateMultiset(e));

        Result<Table1Record> result = ctx.fetchValue(ms);
        assertEquals(resultOne, result);
    }
}
```

The problem here is that the `ResultSet.getString()` method returns the `Result.format()` content:

```
+----+-----+------+
| ID1|NAME1|DATE1 |
+----+-----+------+
|   1|1    |{null}|
+----+-----+------+
```

Instead of the XML or JSON content, depending on `Settings.emulateMultiset`

**Comment by lukaseder** [comment]
The underlying problem is that `ConverterProvider` can't seem to convert from `Record` or `Result` to `String` in the most appropriate way.

**Comment by lukaseder** [comment]
A reminder has been added to https://github.com/jOOQ/jOOQ/issues/11824 about mocking also nested `ROW` types, not just `MULTISET`.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.16.0 and 3.15.3 (https://github.com/jOOQ/jOOQ/issues/12402). Thanks again for your report.
